pkg/agent/corestore.go

```diff
@@ -0,0 +1,127 @@
+package agent
+
+import (
+	"fmt"
+
+	"github.com/datawire/ambassador/pkg/kates"
+	v1 "k8s.io/api/core/v1"
+)
+
+const (
+	NamespaceKubeSystem = "kube-system"
+)
+
+// CoreSnapshot reflects the current state of the kates accumulators for the given
+// resource types.
+type CoreSnapshot struct {
+	Pods        []*kates.Pod
+	ConfigMaps  []*kates.ConfigMap
+	Deployments []*kates.Deployment
+}
+
+// coreStore is used to store core k8s resources that are not handled by default
+// by Ambassador's snapshot.
+type coreStore struct {
+	podStore        *podStore
+	configMapStore  *configMapStore
+	deploymentStore *deploymentStore
+}
+
+type configMapStore struct {
+	sotw map[string]*kates.ConfigMap
+}
+
+type deploymentStore struct {
+	sotw map[string]*kates.Deployment
+}
+
+type podStore struct {
+	sotw map[string]*kates.Pod
+}
+
+// NewCoreStore will create a new coreStore with the given coreSnapshot.
+func NewCoreStore(snapshot *CoreSnapshot) *coreStore {
+	return &coreStore{
+		podStore:        NewPodStore(snapshot.Pods),
+		configMapStore:  NewConfigMapStore(snapshot.ConfigMaps),
+		deploymentStore: NewDeploymentStore(snapshot.Deployments),
+	}
+}
+
+// NewPodStore will create a new podStore filtering out undesired resources.
+func NewPodStore(pods []*kates.Pod) *podStore {
+	sotw := make(map[string]*kates.Pod)
+	store := &podStore{sotw: sotw}
+
+	for _, pod := range pods {
+		if allowedNamespace(pod.GetNamespace()) && pod.Status.Phase != v1.PodSucceeded {
```

Comment:
does it also make sense to filter out failed here?

Reply:
We also spoke about this and we found useful to send pods that are in failed state so we could inspect them and provide useful info to our users on how to address those issues.